File: bin/novel.rb
#!/usr/bin/env ruby
# coding: utf-8
#
require 'yaml'
require 'net/http'
require 'uri'
require 'iconv'

dbfile = File.join(ENV['HOME'],'.backup','latest.yml')
bookid = {'6507'=>'卡徒',
        #'8823'=>'仙道求索',
        '18487'=>'猎国',
        #'10975'=>'天王',
        #'17062'=>'斗罗大陆',
        '17909' => '步步生莲',
        '19116' => '夔牛记',
        #'14059' => '阳神',
        '18663'=>'阴阳冕',
        '18785'=>'仙葫',
        #'405'=>'超级骷髅兵'
        }

def getLatestNovel(url)
    page = Net::HTTP.get(URI.parse(url))
    page = Iconv.conv('utf8//ignore','gbk//ignore',page)
    matches = page.scan(/<a href="(\d+\.html)">(.*?)<\/a>/)
    return matches
end

if __FILE__==$0
    Thread.abort_on_exception = true
    threads = []
    alist = {}
    url0 = "http://www.3jzw.com/files/article/html/%d/%s/"
    bookid.each_key do |key|
        threads << Thread.new { 
            begin
                url =  url0 % [key.to_i/1000, key] + "index.html"
                alist[key] = getLatestNovel(url) 
            rescue Errno::ETIMEDOUT, Timeout::Error, Errno::ECONNRESET
                #$stderr.puts "Connection timed out!"
                exit
            rescue SocketError
                #$stderr.puts "Connection problem, check you internet connection!"
                exit
            end
        }
    end
    threads.each {|t| t.join}

    #f = File.open(dbfile,'r+')
    latest = (File.zero?(dbfile) ? {} : YAML.load_file(dbfile))
    alist.each_key do |key|
        next if alist[key].empty?  

        if latest.key?( key )
            #屏幕输出更新
            oldi = alist[key].find_index {|x| x[0] == latest[key] }
            if oldi != nil and oldi < alist[key].length - 1
                puts "\e[34;1m#{bookid[key]}\e[m"
                alist[key][oldi+1..-1].each { |item| 
                    puts "\t#{item[1]}"
                    puts "\t%s" % (url0 % [key.to_i/1000,key] + item[0]) 
                }
            elsif oldi == nil and not alist[key].empty?
                puts "\e[34;1m#{bookid[key]}\e[m has been updated!"
                puts "\t#{alist[key][-1][1]}"
                puts "\t%s" % (url0 % [key.to_i/1000,key] + alist[key][-1][0]) 
            end
        end
        #初始化数据库
        latest[key] = alist[key][-1][0]
    end
    open(dbfile,'w') { |f| f.puts latest.to_yaml }
end

File: bin/novel.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#Author: Roy L Zuo (roylzuo at gmail dot com)
#Last Change: Fri Mar 13 10:12:31 2009 EST
#Description: 
import urllib2, re, shelve, os, sys
try:
    import pysco
    pysco.full()
except:
    pass

ixxwids = [('卡徒','6507'),
        ('盘龙','4191'), 
        #('巫颂','74029'), 
        ('恶魔法则','150'),
        ('天王','10975'),
        ('蔚蓝轨迹','6530'),
        ('斗罗大陆','10072'),
        ('超级骷髅兵','405')]

def getLatestNovel(url, name, last):
    page = urllib2.urlopen(url).read().decode('gbk').encode('utf8')
    matches = re.findall('<a href="(\d+\.html)">(.*?)</a>', page)
    if last:
        lastindex = 0
        for item in matches:
            if item[0] == last:      break
            lastindex += 1
        if lastindex == len(matches)-1:      return []
        if lastindex > len(matches)-1:      return [matches[-1]]
        return matches[lastindex+1:]
    else:
        return matches

if __name__=="__main__":
    latest = shelve.open("%s/.backup/latestchap" %os.environ['HOME'])
    for item in ixxwids:
        url = "http://www.3jzw.com/files/article/html/%d/%s/index.html" \
                %(int(item[1])/1000, item[1]) 
        try:
            if latest.has_key(item[1]):
                alist = getLatestNovel(url, item[0], latest[item[1]])
                if alist:
                    latest[item[1]] = alist[-1][0]
                    print "\033[34;1m%s\033[m" %item[0]
                    for j in alist:
                        print "\t%s" %j[1]
                        print "\t%s" %url.replace("index.html", j[0])
            else:
                #rebuild the database if the links have been modified
                alist = getLatestNovel(url, item[0], None)
                latest[item[1]] = alist[-1][0]
        except:
            pass
